perceiver_pytorch/encoders.py

```diff
@@ -0,0 +1,353 @@
+import torch
+import torch.nn.functional as F
+import numpy as np
+import math
+import einops
+
+def extract_image_patches(x, kernel, stride=1, dilation=1):
+    # Do TF 'SAME' Padding
+    b, c, h, w = x.shape
+    h2 = math.ceil(h / stride)
+    w2 = math.ceil(w / stride)
+    pad_row = (h2 - 1) * stride + (kernel - 1) * dilation + 1 - h
+    pad_col = (w2 - 1) * stride + (kernel - 1) * dilation + 1 - w
+    x = F.pad(x, (pad_row // 2, pad_row - pad_row // 2, pad_col // 2, pad_col - pad_col // 2))
+
+    # Extract patches
+    patches = x.unfold(2, kernel, stride).unfold(3, kernel, stride)
+    patches = patches.permute(0, 4, 5, 1, 2, 3).contiguous()
+
+    return patches.view(b, -1, patches.shape[-2], patches.shape[-1])
+
+
+def reverse_space_to_depth(
+        frames: np.ndarray, temporal_block_size: int = 1, spatial_block_size: int = 1
+) -> np.ndarray:
+    """Reverse space to depth transform."""
+    if len(frames.shape) == 4:
+        return einops.rearrange(
+            frames,
+            "b h w (dh dw c) -> b (h dh) (w dw) c",
+            dh=spatial_block_size,
+            dw=spatial_block_size,
+        )
+    elif len(frames.shape) == 5:
+        return einops.rearrange(
+            frames,
+            "b t h w (dt dh dw c) -> b (t dt) (h dh) (w dw) c",
+            dt=temporal_block_size,
+            dh=spatial_block_size,
+            dw=spatial_block_size,
+        )
+    else:
+        raise ValueError(
+            "Frames should be of rank 4 (batch, height, width, channels)"
+            " or rank 5 (batch, time, height, width, channels)"
+        )
+
+
+def space_to_depth(
+        frames: np.ndarray, temporal_block_size: int = 1, spatial_block_size: int = 1
+) -> np.ndarray:
+    """Space to depth transform."""
+    if len(frames.shape) == 4:
+        return einops.rearrange(
+            frames,
+            "b (h dh) (w dw) c -> b h w (dh dw c)",
+            dh=spatial_block_size,
+            dw=spatial_block_size,
+        )
+    elif len(frames.shape) == 5:
+        return einops.rearrange(
+            frames,
+            "b (t dt) (h dh) (w dw) c -> b t h w (dt dh dw c)",
+            dt=temporal_block_size,
+            dh=spatial_block_size,
+            dw=spatial_block_size,
+        )
+    else:
+        raise ValueError(
+            "Frames should be of rank 4 (batch, height, width, channels)"
+            " or rank 5 (batch, time, height, width, channels)"
+        )
+
+
+class Conv2DDownsample(torch.nn.Sequential):
```

Comment:
Should these functions be a separate file called 'convolution'?

Reply:
Done!